File: kruskal.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <map>

// Estrutura para representar uma aresta:
struct Aresta {
    std::string origem,destino;
    int peso;
};

// Função para comparar arestas com base em seus pesos:
bool compararArestas(const Aresta& a, const Aresta& b) {
    return a.peso < b.peso;
}

// Função para encontrar a raiz de um conjunto usando o algoritmo de busca em árvore:
std::string  encontrarRaiz(std::string vertice, std::map<std::string, std::string>& conjunto) {
    while (conjunto[vertice] != vertice) {
        vertice = conjunto[vertice];
    }
    return vertice;
}

// Função para unir dois subconjuntos em um conjunto maior:
void juntarArestas(std::string raizA, std::string raizB, std::map<std::string, std::string>& conjunto) {
    conjunto[raizB] = raizA;
}

// Algoritmo de Kruskal:
void kruskal(std::vector<Aresta>& grafo) {
    // Ordena as arestas em ordem crescente de peso
    sort(grafo.begin(), grafo.end(), compararArestas);

    std::map<std::string, std::string> conjunto;

    // Adiciona ao conjunto os pares de origem e destino
    for (const Aresta& aresta : grafo) {
        conjunto[aresta.origem] = aresta.origem;
        conjunto[aresta.destino] = aresta.destino;
    }

    std::vector<Aresta> arvoreMinima;

    // checa se houve ciclo 
    for (const Aresta& aresta : grafo) {
        std::string raizOrigem = encontrarRaiz(aresta.origem, conjunto);
        std::string raizDestino = encontrarRaiz(aresta.destino, conjunto);

        // Se a raiz de origem e a raiz de destino forem diferentes, a aresta pode ser adicionada à árvore mínima
        if (raizOrigem != raizDestino) {
            arvoreMinima.push_back(aresta);
            juntarArestas(raizOrigem, raizDestino, conjunto); // assim o conjunto "liga as arestas ja visitadas"
        }
    }

    // Imprime árvore mínima
    for (const Aresta& aresta : arvoreMinima) {
        std::cout << aresta.origem << " - " << aresta.destino << " : " << aresta.peso << std::endl;
    }
}




int main() {
    std::vector<Aresta> grafo;

    grafo.push_back({"A","B", 2});
    grafo.push_back({"A","C", 3});
    grafo.push_back({"A","D", 3});
    grafo.push_back({"B","C", 4});
    grafo.push_back({"B","E", 3});
    grafo.push_back({"C","D", 5});
    grafo.push_back({"C","E", 1});
    grafo.push_back({"D","F", 7});
    grafo.push_back({"E","F", 8});
    grafo.push_back({"F","G", 9});

    kruskal(grafo);

    return 0;
}

File: Kruskal.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

// Classe para representar uma aresta:
class Aresta {
    String origem, destino;
    int peso;

    public Aresta(String origem, String destino, int peso) {
        this.origem = origem;
        this.destino = destino;
        this.peso = peso;
    }
}

public class Kruskal {

    // Função para comparar arestas com base em seus pesos:
    static class ComparadorArestas implements java.util.Comparator<Aresta> {
        @Override
        public int compare(Aresta a, Aresta b) {
            return Integer.compare(a.peso, b.peso);
        }
    }

    // Função para encontrar a raiz de um conjunto usando o algoritmo de busca em árvore:
    static String encontrarRaiz(String vertice, Map<String, String> conjunto) {
        while (!conjunto.get(vertice).equals(vertice)) {
            vertice = conjunto.get(vertice);
        }
        return vertice;
    }

    // Função para unir dois subconjuntos em um conjunto maior:
    static void juntarArestas(String raizA, String raizB, Map<String, String> conjunto) {
        conjunto.put(raizB, raizA);
    }

    // Algoritmo de Kruskal:
    static void kruskal(List<Aresta> grafo) {
        // Ordena as arestas em ordem crescente de peso
        Collections.sort(grafo, new ComparadorArestas());

        Map<String, String> conjunto = new HashMap<>();

        // Adiciona ao conjunto os pares de origem e destino
        for (Aresta aresta : grafo) {
            conjunto.putIfAbsent(aresta.origem, aresta.origem);
            conjunto.putIfAbsent(aresta.destino, aresta.destino);
        }

        List<Aresta> arvoreMinima = new ArrayList<>();

        // checa se houve ciclo
        for (Aresta aresta : grafo) {
            String raizOrigem = encontrarRaiz(aresta.origem, conjunto);
            String raizDestino = encontrarRaiz(aresta.destino, conjunto);

            // Se a raiz de origem e a raiz de destino forem diferentes, a aresta pode ser adicionada à árvore mínima
            if (!raizOrigem.equals(raizDestino)) {
                arvoreMinima.add(aresta);
                juntarArestas(raizOrigem, raizDestino, conjunto); // assim o conjunto "liga as arestas ja visitadas"
            }
        }

        // Imprime árvore mínima
        for (Aresta aresta : arvoreMinima) {
            System.out.println(aresta.origem + " - " + aresta.destino + " : " + aresta.peso);
        }
    }

    public static void main(String[] args) {
        List<Aresta> grafo = new ArrayList<>();

        grafo.add(new Aresta("A", "B", 2));
        grafo.add(new Aresta("A", "C", 3));
        grafo.add(new Aresta("A", "D", 3));
        grafo.add(new Aresta("B", "C", 4));
        grafo.add(new Aresta("B", "E", 3));
        grafo.add(new Aresta("C", "D", 5));
        grafo.add(new Aresta("C", "E", 1));
        grafo.add(new Aresta("D", "F", 7));
        grafo.add(new Aresta("E", "F", 8));
        grafo.add(new Aresta("F", "G", 9));

        kruskal(grafo);
    }
}
